src/main/java/org/openjdk/jextract/JextractTool.java

```diff
@@ -115,34 +103,34 @@ private static Path generateTmpSource(List<Path> headers) {
      */
     public static Declaration.Scoped parse(List<Path> headers, String... parserOptions) {
+        return parseInternal(Logger.DEFAULT, headers, parserOptions);
+    }
+    private static Declaration.Scoped parseInternal(Logger logger, List<Path> headers, String... parserOptions) {
         Path source = headers.size() > 1? generateTmpSource(headers) : headers.iterator().next();
-        return new Parser().parse(source, Stream.of(parserOptions).collect(Collectors.toList()));
+        return new Parser(logger)
+                .parse(source, Stream.of(parserOptions).collect(Collectors.toList()));
     }
 
     public static List<JavaSourceFile> generate(Declaration.Scoped decl, String headerName,
                                                 String targetPkg, List<Options.Library> libs,
-                                                boolean useSystemLoadLibrary, PrintWriter errStream) {
-        return List.of(generate(decl, headerName, targetPkg, new IncludeHelper(), libs, useSystemLoadLibrary, errStream));
+                                                boolean useSystemLoadLibrary) {
+        return generateInternal(decl, headerName, targetPkg, new IncludeHelper(),
+                libs, useSystemLoadLibrary, Logger.DEFAULT);
     }
 
     private static List<JavaSourceFile> generateInternal(Declaration.Scoped decl, String headerName,
                                                          String targetPkg, IncludeHelper includeHelper,
                                                          List<Options.Library> libs, boolean useSystemLoadLibrary,
-                                                         PrintWriter errStream) {
-        return List.of(generate(decl, headerName, targetPkg, includeHelper, libs, useSystemLoadLibrary, errStream));
-    }
-
-    private static JavaSourceFile[] generate(Declaration.Scoped decl, String headerName,
-                                             String targetPkg, IncludeHelper includeHelper,
-                                             List<Options.Library> libs, boolean useSystemLoadLibrary,
-                                             PrintWriter errStream) {
+                                                         Logger logger) {
         var transformedDecl = Stream.of(decl)
-                .map(new IncludeFilter(includeHelper)::scan)
+                .map(new IncludeFilter(includeHelper, logger)::scan)
                 .map(new DuplicateFilter()::scan)
                 .map(new NameMangler(headerName)::scan)
-                .map(new UnsupportedFilter(errStream)::scan)
+                .map(new UnsupportedFilter(logger)::scan)
                 .findFirst().get();
-        return OutputFactory.generateWrapped(transformedDecl, targetPkg, libs, useSystemLoadLibrary);
+        return logger.hasErrors() ?
+                List.of() :
+                List.of(OutputFactory.generateWrapped(transformedDecl, targetPkg, libs, useSystemLoadLibrary));
```

Comment:
So, we now don't generate anything when there are errors.

I think we should also return a non-zero exit code in that case.

Reply:
True